infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -41,35 +178,77 @@ public String getName() {
         return "RestoreRedundancyAndMergeSegments";
     }
 
+    @FunctionalInterface
+    public interface LogUnitClientMap {
+        Map<String, LogUnitClient> generate(CorfuRuntime runtime);
+    }
+
     @Override
     public void impl(@Nonnull CorfuRuntime runtime) throws Exception {
 
         // Refresh layout.
         runtime.invalidateLayout();
         Layout layout = runtime.getLayoutView().getLayout();
 
-        // Each segment is compared with the first segment. The data is restored in any new LogUnit nodes and then
-        // merged to this segment. This is done for all the segments.
-        final int layoutSegmentToMergeTo = 0;
+        // Create a helper to perform the state calculations.
+        PrefixTrimRedundancyCalculator redundancyCalculator =
+                new PrefixTrimRedundancyCalculator(currentNode, runtime);
 
-        // Catchup all servers across all segments.
-        while (layout.getSegments().size() > 1) {
+        // Create a state transfer manager.
+        StateTransferManager transferManager =
+                new StateTransferManager(streamLog, runtime.getParameters().getBulkReadSize());
 
-            Set<String> lowRedundancyServers = getNodesWithReducedRedundancy(layout, layoutSegmentToMergeTo);
+        // Create lambda that, given a runtime, creates the log unit clients map.
+        LogUnitClientMap map = (CorfuRuntime rt) -> {
+            Layout currentLayout = rt.getLayoutView().getLayout();
+            Set<String> allActiveServers = currentLayout.getAllActiveServers();
 
-            // Currently the state is transferred for the complete segment.
-            // TODO: Add stripe specific transfer granularity for optimization.
-            // Transfer the replicated segment to the difference set calculated above.
-            for (String lowRedundancyServer : lowRedundancyServers) {
-                StateTransfer.transfer(layout, lowRedundancyServer, runtime, layout.getFirstSegment());
-            }
+            return allActiveServers.stream()
+                    .map(server -> {
+                        RuntimeLayout runtimeLayout = rt.getLayoutView().getRuntimeLayout();
+                        return new Tuple<>(server, runtimeLayout.getLogUnitClient(server));
+                    })
+                    .collect(Collectors.toMap(tuple -> tuple.first, tuple -> tuple.second));
+        };
+
+        // Create the initial state map.
+        ImmutableList<TransferSegment> stateList =
+                redundancyCalculator.createStateList(layout);
+
+        log.info("State transfer on {}: Initial state list: {}", currentNode, stateList);
+
+        while (RedundancyCalculator.requiresRedundancyRestoration(layout, currentNode) ||
```

Comment:
you can provide an additional method in RedundancyCalculator that check both conditions.
Like `RedundancyCalculator.needsRestorationOrMerge`

Reply:
Fixed